### PR TITLE
Refactor OpenAPI Generator

### PR DESCRIPTION
**Pull request by samchungy**
Was trying to unwrap some of the code, looks like we have some duplicate code which doesn't do anything.

This will help with https://github.com/asteasolutions/zod-to-openapi/issues/99#issue-1604769637

### CODE REVIEW

File: src/openapi-generator.ts [hunk 1]
```diff
@@ -333,7 +333,7 @@ export class OpenAPIGenerator {
     const required =
       !this.isOptionalSchema(zodSchema) && !zodSchema.isNullable();
 
-    const schema = this.generateSimpleSchema(zodSchema);
+    const schema = this.generateSchemaDefinition(zodSchema);
```
Comment (AGalabov) [inline src/openapi-generator.ts:336]: I don't think I can agree with this one. Right now this change would enforce a schema being defined into `schemaRefs` and would then be saved in `paramRefs` (in `generateParameterDefinition`). That was the overall idea behind having a `generateSimpleSchema` that does not bother "saving the result ref".

File: src/openapi-generator.ts [hunk 2]
```diff
@@ -1113,12 +1098,12 @@ export class OpenAPIGenerator {
 
   private applySchemaMetadata(
     initialData: SchemaObject | ParameterObject | ReferenceObject,
-    metadata: Partial<ZodOpenAPIMetadata>
+    metadata?: Partial<ZodOpenAPIMetadata>
```
Comment (AGalabov) [inline src/openapi-generator.ts:1101]: I am not a big fan of this even if it helps with some of the code.

Basically right now we have a method saying "apply metadata" that can work without metadata - that's very odd in my opinion. I guess we can work around this by renaming the method :thinking:. Not sure what could be a suitable name.

And one more thing - if we do end up renaming it and sticking with the optional, there are a couple more instances where a ternary operator was used to cover for undefined metadata.

**Review by AGalabov — COMMENTED** [review]
@samchungy I get the idea and deep down I know that there is some duplication so I am very much up for a refactoring. However, I did raise some questions. And additionally can you share how you envision this helping with #99?

**Comment by samchungy** [comment]
> @samchungy I get the idea and deep down I know that there is some duplication so I am very much up for a refactoring. However, I did raise some questions. And additionally can you share how you envision this helping with #99?

How do I register a schema without calling `.register`? I'm thinking that, while we traverse the Zod object - whenever we run into `refId` or `schemaRef` or TBD, we add it to `this.schemaRefs` and create a ref to it. In my head I think that makes sense

File: src/openapi-generator.ts (hunk 1 above)
Comment (samchungy) [reply src/openapi-generator.ts:336]: Oh I see I see, let me rethink this :)

**Pull request closed without merging**